aggregator/operator_registrations_inmemory.go

```diff
@@ -0,0 +1,190 @@
+	respC chan<- resp
+}
+
+type resp struct {
+	operatorPubkeys types.OperatorPubkeys
+	// false if operators were not present in the pubkey dict
+	operatorExists bool
+}
+
+var _ operatorpubkeys.OperatorPubkeysService = (*OperatorRegistrationsServiceInMemory)(nil)
+
+// NewOperatorRegistrationsServiceInMemory constructs a OperatorRegistrationsServiceInMemory and starts it in a goroutine.
+// It takes a context as argument because the "backfilling" of the database is done inside this constructor,
+// so we wait for all past NewPubkeyRegistration events to be queried and the db to be filled before returning the service.
+// The constructor is thus following a RAII-like pattern, of initializing the serving during construction.
+// Using a separate initialize() function might lead to some users forgetting to call it and the service not behaving properly.
+func NewOperatorRegistrationsServiceInMemory(
+	ctx context.Context,
+	avsRegistrySubscriber avsregistry.AvsRegistrySubscriber,
+	avsRegistryReader avsregistry.AvsRegistryReader,
+	logger logging.Logger,
+) *OperatorRegistrationsServiceInMemory {
+	queryByAddrC := make(chan queryByAddr)
+	queryByIdC := make(chan queryById)
+
+	pkcs := &OperatorRegistrationsServiceInMemory{
+		avsRegistrySubscriber: avsRegistrySubscriber,
+		avsRegistryReader:     avsRegistryReader,
+		logger:                logger,
+		queryByAddrC:          queryByAddrC,
+		queryByIdC:            queryByIdC,
+	}
+
+	// We use this waitgroup to wait on the initialization of the inmemory pubkey dict,
+	// which requires querying the past events of the pubkey registration contract
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	pkcs.startServiceInGoroutine(ctx, queryByAddrC, queryByIdC, &wg)
+	wg.Wait()
+
+	return pkcs
+}
+
+func (ors *OperatorRegistrationsServiceInMemory) startServiceInGoroutine(ctx context.Context, queryByAddrC <-chan queryByAddr, queryByIdC <-chan queryById, wg *sync.WaitGroup) {
+	go func() {
+		pubkeyByAddrDict := make(map[common.Address]types.OperatorPubkeys)
+		pubkeyByIdDict := make(map[types.OperatorId]types.OperatorPubkeys)
+
+		ors.logger.Debug("Subscribing to new pubkey registration events on blsApkRegistry contract", "service", "OperatorRegistrationsServiceInMemory")
+		newPubkeyRegistrationC, newPubkeyRegistrationSub, err := ors.avsRegistrySubscriber.SubscribeToNewPubkeyRegistrations()
+		if err != nil {
+			ors.logger.Error("Fatal error opening websocket subscription for new pubkey registrations", "err", err, "service", "OperatorRegistrationsServiceInMemory")
+			// see the warning above the struct definition to understand why we panic here
+			panic(err)
+		}
+
+		ors.queryPastRegisteredOperatorEventsAndFillDb(ctx, pubkeyByAddrDict, pubkeyByIdDict)
+
+		// The constructor can return after we have backfilled the db by querying the events of operators that have registered with the blsApkRegistry
+		// before the block at which we started the ws subscription above
+		wg.Done()
```

Comment:
What is the rationale for handling the dict in the service? Overall it seems better to just initialize it directly in the constructor and trigger the 'task' later, it's already waiting anyway. Would avoid the panic as well.

Reply:
https://github.com/NethermindEth/near-sffl/pull/194#discussion_r1620326046

I agree with initial implementation here. Right now constructor look like
`async func NewOperatorRegistrationsServiceInMemory.` What you proposing is 
`func NewOperatorRegistrationsServiceInMemory`. 

In first case you delegate task to scheduler in second you say: I will filter all events right now disregarding all pending tasks in queue.  As developer calling a constructor that does intensive computation is an unexpected surprise. It will do it later anyway but when - is scheduler business.

Imagine a React.JS where while a scheduled task is rerender you start queuing whole history blocking it.

Reply:
I agree in a sense, but the constructor is waiting for it anyway

Reply:
In the end it's like the discussion outcome in the linked comment - for the caller it doesn't really change the situation much, the work is being done in another goroutine but it's being waited anyway. IMO if it's being waited I don't think we have much of a use for doing it in a goroutine, especially with the error handling considerations

Reply:
> I agree in a sense, but the constructor is waiting for it anyway

Exactly, caller is parked and that enables scheduler to progress with other tasks: such as render(in React) or any other scheduled tasks. Intensive work shall be scheduled.

>  for the caller it doesn't really change situation much

It's not only about the current caller. If that wouldn't change anything such things as schedulers in OSs, languages wouldn't exist and everything could just be executed on the thread without asynchronous tasks. 

Imagine a single process running blocking keyboard inputs/SSD and other drivers signals, because it doesn't want to let other tasks to be scheduled.

Reply:
Alright, it seemed to me you hadn't noticed the `wg.Done()` somehow. I'm not disagreeing with the approach specifically, just IMO we still don't really have a use for this extra step in the context we're using this module. If we are doing it like that though, then we should not just throw panics around and we should also do it the 'right way' and e.g. have an error channel or such that is read in the constructor to relay the error.

Reply:
With the way you were describing before, I wasn't sure you were aware it was waiting for the async task anyway. That said, if you're removing the panics and catching the error, that's fine by me

Reply:
Do we need wg considering we are waiting for the error channel anyway?